### PR TITLE
refactor transports to be fully featured

### PR DESCRIPTION
**Pull request by Stebalien**
DO NOT MERGE (yet)
Part of: libp2p/go-libp2p#297

### CODE REVIEW

File: fallback.go [hunk 1]
```diff
@@ -1,53 +0,0 @@
-package transport
```
Comment (Stebalien) [inline fallback.go:1]: Removed without replacement (the libp2p constructor will add-in the tcp transport by default if no other transports are specified.

File: transport.go [hunk 2]
```diff
@@ -24,19 +40,27 @@ type Conn interface {
...
-// that we dial out from.
-type Dialer interface {
-	Dial(raddr ma.Multiaddr) (Conn, error)
-	DialContext(ctx context.Context, raddr ma.Multiaddr) (Conn, error)
-	Matches(ma.Multiaddr) bool
+	// Protocol returns the set of protocols handled by this transport.
+	Protocols() []int
+
+	// Proxy returns true if this transport proxies.
+	// TODO: Make this a part of the go-multiaddr protocol instead?
```
Comment (Stebalien) [inline transport.go]: I think it's fine to leave this here for now but we may want to move it later. Proxy transports are transports like `p2p-circuit` that handle part of the multiaddr and forward the rest to some other system.

File: transport.go [hunk 3]
```diff
@@ -50,9 +74,10 @@ type Listener interface {
 	Multiaddr() ma.Multiaddr
 }
 
-// DialOpt is an option used for configuring dialer behaviour
-type DialOpt interface{}
```
Comment (Stebalien) [inline transport.go:54]: All configuration options are passed to the transport on construction.

File: transport.go [hunk 4]
```diff
@@ -24,19 +40,27 @@ type Conn interface {
 // but many more can be implemented, sctp, audio signals, sneakernet, UDT, a
 // network of drones carrying usb flash drives, and so on.
 type Transport interface {
-	Dialer(laddr ma.Multiaddr, opts ...DialOpt) (Dialer, error)
```
Comment (Stebalien) [inline transport.go:27]: We no longer have "dialers". Instead, we just call `Dial` on the transport and the transport decides how to dial (e.g., which source address to use). In the past, we constructed one dialer per listen address and picked the dialer randomly. That was obviously *not* an optimal solution but the swarm just didn't have enough information to make an informed decision (without special-casing transports). On the other hand, the transport *does* have enough information.

File: transport.go [hunk 5]
```diff
@@ -24,19 +40,27 @@ type Conn interface {
...
+	// Dial dials a remote peer. It should try to reuse local listener
+	// addresses if possible but it may choose not to.
+	Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (Conn, error)
```
Comment (Stebalien) [inline transport.go:45]: Given that we're breaking things, I got rid of the distinction between `Dial` and `DialContext`. Just pass a context.

Also, this now takes a peer ID. While we can technically get away with not having this, it allows us to validate connections early (given that transport security now happens *inside* the transport).

File: README.md [hunk 6]
```diff
@@ -21,26 +21,6 @@ This is the 'base' layer for any transport that wants to be used by libp2p and i
 > gx-go rewrite
 ```
 
-## Usage
```
Comment (whyrusleeping) [inline README.md]: no new usage text? @Stebalien tsk tsk tsk

**Review by whyrusleeping — APPROVED** [review]
LGTM, doing away with the fallback dialer seems okay.

File: README.md (hunk 6 above)
Comment (Stebalien) [reply README.md]: One generally doesn't *use* transports directly like this. One implements them.

... fine. I'll add an example.

**Comment by Stebalien** [comment]
*edit: punt, see #30*

---

So, during the refactor, I inverted the transport/net dependency. That is, instead of go-libp2p-net depending on go-libp2p-transport, go-libp2p-transport now depends on go-libp2p-net. My reasoning was that transports are *one* way to implement a libp2p networks. That's why this package has a `transport.Network` interface that extends `inet.Network`.

However, go-libp2p-net contains the `Network` interface which, unfortunately, depends on the peerstore. Given that all transports will depend on go-libp2p-transport, this is a bit unfortunate.

One solution is to:

1. Move `inet.ConnMultiaddrs` to `go-multiaddr-net`. Really, that's where it belongs (and then `maddr.Conn` would simply be a composition of `net.Conn` and this interface).
2. Move `inet.ConnSecurity` somewhere... Honestly, I don't know where it belongs. Maybe `go-libp2p-peer` or a new `go-libp2p-peer-net` package? We don't really *need* this mixin but it's nice to have.
3. Make a new `go-libp2p-transport-net` package and move the `transport.Network` interface there.

Personally, I'm inclined to finish this refactor and punt on this. Unlike the current refactor, moving these interfaces around won't be a major operation.

**Pull request merged**